### PR TITLE
[FEAT] [ROCm] Upgrade AITER Fused MoE kernels.

### PR DESCRIPTION
**Pull request by vllmellm**
This PR upgrades the AITER fused MoE kernels which introduces a new external API, `fused_moe`, from the `aiter` package. The `fused_moe` API automatically selects and invokes the appropriate type of MoE kernel based on the specified quantization method.

### Qwen/Qwen3-30B-A3B-FP8

`VLLM_USE_V1=1
VLLM_ROCM_USE_AITER=1
VLLM_ROCM_USE_AITER_MOE=1
SAFETENSORS_FAST_GPU=1
lm_eval --model vllm --model_args pretrained=Qwen/Qwen3-30B-A3B-FP8,tensor_parallel_size=1,max_model_len=32768,quantization=fp8,kv_cache_dtype=fp8 --tasks gsm8k --num_fewshot 5 --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8309|±  |0.0103|
|     |       |strict-match    |     5|exact_match|↑  |0.8931|±  |0.0085|

### mistralai/Mixtral-8x7B-Instruct-v0.1 no quantization

`VLLM_USE_V1=1
VLLM_ROCM_USE_AITER=1
SAFETENSORS_FAST_GPU=1
lm_eval --model vllm --model_args pretrained=mistralai/Mixtral-8x7B-Instruct-v0.1,tensor_parallel_size=2,max_model_len=10000 --tasks gsm8k  --num_fewshot 5 --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6437|±  |0.0132|
|     |       |strict-match    |     5|exact_match|↑  |0.6399|±  |0.0132|

### mistralai/Mixtral-8x7B-Instruct-v0.1 fp8 quantization

`VLLM_USE_V1=1
VLLM_ROCM_USE_AITER=1
SAFETENSORS_FAST_GPU=1
lm_eval --model vllm --model_args pretrained=mistralai/Mixtral-8x7B-Instruct-v0.1,tensor_parallel_size=2,max_model_len=10000,quantization=fp8 --tasks gsm8k  --num_fewshot 5 --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5474|±  |0.0137|
|     |       |strict-match    |     5|exact_match|↑  |0.5353|±  |0.0137|

### deepseek-ai/DeepSeek-V3 on v0 engine
`VLLM_USE_V1=0
VLLM_ROCM_USE_AITER=1
SAFETENSORS_FAST_GPU=1
lm_eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V3,tensor_parallel_size=8,max_model_len=32768,block_size=1 --trust_remote_code --tasks gsm8k --num_fewshot 5 --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9249|±  |0.0073|
|     |       |strict-match    |     5|exact_match|↑  |0.9242|±  |0.0073|

### RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic
`VLLM_USE_V1=1
VLLM_ROCM_USE_AITER=1
VLLM_ROCM_USE_AITER_RMSNORM=0
SAFETENSORS_FAST_GPU=1
lm_eval --model vllm --model_args pretrained=RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic,gpu_memory_utilization=0.95,tensor_parallel_size=8,max_model_len=32768 --tasks gsm8k --num_fewshot 5 --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9212|±  |0.0074|
|     |       |strict-match    |     5|exact_match|↑  |0.9075|±  |0.0080|

### meta-llama/Llama-4-Scout-17B-16E-Instruct
`VLLM_USE_V1=1 
VLLM_ROCM_USE_AITER=1 
VLLM_ROCM_USE_AITER_RMSNORM=0 
SAFETENSORS_FAST_GPU=1 
lm_eval --model vllm --model_args pretrained=meta-llama/Llama-4-Scout-17B-16E-Instruct,gpu_memory_utilization=0.95,tensor_parallel_size=8,max_model_len=32768 --tasks gsm8k --num_fewshot 5 --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9189|±  |0.0075|
|     |       |strict-match    |     5|exact_match|↑  |0.9052|±  |0.0081|

## Benchmarking Results

| Metric                        | Mixtral-8x7B-v0.1, fp16, New | Mixtral-8x7B-v0.1, fp8, New | Mixtral-8x7B-v0.1, fp16, Old | Mixtral-8x7B-v0.1, fp8, Old |
|-------------------------------|------------------------------|-----------------------------|------------------------------|-----------------------------|
| **AITER API Commit**          | New (c1debd8)                | New (c1debd8)               | Old (5a77249)                | Old (5a77249)               |
| **Benchmark Duration (s)**    | 162.58                       | 29.31                       | 188.76                       | 122.67                      |
| **Successful Requests**       | 500                          | 500                         | 500                          | 500                         |
| **Input Tokens**              | 493,277                      | 493,277                     | 493,277                      | 493,277                     |
| **Output Tokens**             | 456,279                      | 57,752                      | 449,576                      | 439,241                     |
| **Request Throughput (req/s)**| 3.08                         | 17.06                       | 2.65                         | 4.08                        |
| **Goodput (req/s)**           | 0.00                         | 0.00                        | 0.01                         | 0.01                        |
| **Output Token Throughput**   | 2,806.56                     | 1,970.57                    | 2,381.74                     | 3,580.56                    |
| **Total Token Throughput**    | 5,840.70                     | 18,801.83                   | 4,994.99                     | 7,601.60                    |
| **Mean TTFT (ms)**            | 19,469.03                    | 11,732.31                   | 20,959.97                    | 17,232.23                   |
| **Median TTFT (ms)**          | 19,850.54                    | 11,281.37                   | 20,437.42                    | 17,028.30                   |
| **P99 TTFT (ms)**             | 37,288.14                    | 22,810.18                   | 41,393.29                    | 35,066.92                   |
| **Mean TPOT (ms)**            | 142.90                       | 149.83                      | 173.39                       | 136.56                      |
| **Median TPOT (ms)**          | 95.36                        | 144.57                      | 118.59                       | 77.03                       |
| **P99 TPOT (ms)**             | 601.69                       | 375.53                      | 667.27                       | 571.96                      |
| **Mean ITL (ms)**             | 91.96                        | 138.70                      | 111.11                       | 73.88                       |
| **Median ITL (ms)**           | 78.96                        | 65.05                       | 98.11                        | 58.36                       |
| **P99 ITL (ms)**              | 593.61                       | 388.65                      | 662.64                       | 566.49                      |

### CODE REVIEW

**Comment by robertgshaw2-redhat** [comment]
<3 aiter

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -376,10 +376,9 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
             shuffle_weights)
 
         if self.rocm_aiter_moe_enabled:
-            # use 2stage ck moe layout
             shuffled_w13, shuffled_w2 = shuffle_weights(layer.w13_weight.data,
                                                         layer.w2_weight.data,
-                                                        layout=(32, 32))
+                                                        layout=(16, 16))
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/layer.py:381]: can this be made into something static that clarifies what it is (e.g. 
`AITER_XXX=16`). This constant in the middle of the function is just confusing

Comment (vllmellm) [reply vllm/model_executor/layers/fused_moe/layer.py]: @robertgshaw2-redhat updated the documentation for shuffle_weights as below.  which explains and clarifies the arguments of the function. 

https://github.com/EmbeddedLLM/vllm/blob/85a7151d180de411d9593f7831a5f1d8c437685f/vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py#L351-L372

Since currently the best optimal layout size for all kernels are the same as (16, 16) we kept that as default and in future updates if the layout meant to change for any kernel then we can introduce the constant variables based on individual kernels in `rocm_aiter_fused_moe.py` and use them where needed.

File: vllm/model_executor/layers/quantization/fp8.py [hunk 2]
```diff
@@ -762,20 +748,9 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
-                    max_w13_scales,
-                    layer.w2_weight_scale.data,
-                    expansion_dims=expansion_dims)
-                layer.w2_weight_scale = torch.nn.Parameter(
-                    w2_scales.contiguous(), requires_grad=False)
-
                 shuffled_w13, shuffled_w2 = shuffle_weights(layer.w13_weight,
                                                             layer.w2_weight,
-                                                            layout=(32, 32))
+                                                            layout=(16, 16))
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/fp8.py:753]: Nit: Remove `layout` param like elsewhere?

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py [hunk 3]
```diff
@@ -274,6 +140,50 @@ def rocm_aiter_biased_grouped_topk_fake(
...
 
+def rocm_aiter_fused_moe_impl(
+    hidden_states: torch.Tensor,
+    w1: torch.Tensor,
+    w2: torch.Tensor,
+    topk_weight: torch.Tensor,
+    topk_ids: torch.Tensor,
+    expert_mask: Optional[torch.Tensor] = None,
+    activation_method: int = ActivationMethod.SILU.value,
+    quant_method: int = QuantMethod.NO.value,
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py:151]: Why not require the enum to be passed in here, and then do `.value` inside the function?
```suggestion
    activation_method: ActivationMethod = ActivationMethod.SILU,
    quant_method: QuantMethod = QuantMethod.NO,
```

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py [hunk 4]
```diff
@@ -274,6 +140,50 @@ def rocm_aiter_biased_grouped_topk_fake(
...
+    w1_scale: Optional[torch.Tensor] = None,
+    w2_scale: Optional[torch.Tensor] = None,
+    a1_scale: Optional[torch.Tensor] = None,
+    a2_scale: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    from aiter import ActivationType, QuantType
+    from aiter.fused_moe import fused_moe
+
+    activation = ActivationType(activation_method)
+    quant_type = QuantType(quant_method)
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py:162]:
```suggestion
    activation = ActivationType(activation_method.value)
    quant_type = QuantType(quant_method.value)
```

**Review by ProExpertProg — COMMENTED** [review]
Looks pretty good to me but I'll also let @SageMoore review once he's back next week

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py (hunk 3 above)
Comment (vllmellm) [reply vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py:151]: @ProExpertProg unfortunately we can't use the Enum type directly as the `direct_register_custom_op` complains about Enum type argument and does not register the function in torch schema.

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py (hunk 4 above)
Comment (vllmellm) [reply vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py:162]: @ProExpertProg The QuantType from aiter package is not python Enum and it's been defined in a .h file from aiter csrc package and it is used in python using pybind and it's object cannot be created by python enum object as your suggestion here either. It has to be created using its integer value. Therefore the introduced Enum classes introduced in this module as commented in the code is meant to be used for interfacing with the respective enum object from aiter package.

File: vllm/model_executor/layers/quantization/fp8.py (hunk 2 above)
Comment (vllmellm) [reply vllm/model_executor/layers/quantization/fp8.py]: @ProExpertProg Thanks for pointing out. is removed now.

**Review by SageMoore — APPROVED** [review]
Always good to see a PR that deletes more code than it adds :). LGTM

**Review by DarkLight1337 — APPROVED** [review]
Test failure is unrelated, good to merge
